lib/klaviyo/client.rb

```diff
@@ -1,24 +1,185 @@
 require 'open-uri'
 require 'base64'
 require 'json'
+require 'Faraday'
 
 module Klaviyo
   class KlaviyoError < StandardError; end
-  
+
+  NO_PRIVATE_API_KEY_ERROR = 'Please provide Private API key for this request'
+
   class Client
-    def initialize(api_key, url = 'http://a.klaviyo.com/')
+    def initialize(api_key = nil, private_api_key = nil, domain = 'https://a.klaviyo.com/api')
+      if !api_key
+        KlaviyoError.new('Please provide your Public API key')
+      end
+
       @api_key = api_key
-      @url = url
+      @private_api_key = private_api_key
+      @domain = domain
+
+      if @private_api_key
+        @private_api_key_param = "api_key=#{@private_api_key}"
+      end
+
+      @v1 = 'v1'
+      @v2 = 'v2'
+      @metric = 'metric'
+      @metrics = 'metrics'
+      @timeline = 'timeline'
+      @list = 'list'
+      @lists = 'lists'
+      @person = 'person'
+      @subscribe = 'subscribe'
+      @members = 'members'
+      @exclusions = 'exclusions'
+      @all = 'all'
+      @group = 'group'
+
+    end
+
+# METRICS API
+
+    def get_metrics(kwargs = {})
+      path = @metrics
+
+      v1_request('GET', path, kwargs)
+    end
+
+    def get_metrics_timeline(kwargs = {})
+      path = "#{@metrics}/#{@timeline}"
+
+      v1_request('GET', path, kwargs)
+    end
+
+    def get_specific_metric_timeline(metric_id, kwargs = {})
+      path = "#{@metric}/#{metric_id}/#{@timeline}"
+
+      v1_request('GET', path, kwargs)
+    end
+
+# PROFILES API
+
+    def get_person_attributes(person_id)
+      path = "#{@person}/#{person_id}"
+
+      v1_request('GET', path)
+    end
+
+    def update_person_attributes(person_id, kwargs = {})
+      path = "#{@person}/#{person_id}"
+
+      v1_request('PUT', path, kwargs)
+    end
+
+    def get_person_metrics_timeline(person_id, kwargs = {})
+      path = "#{@person}/#{person_id}/#{@metrics}/#{@timeline}"
+
+      v1_request('GET', path, kwargs)
+    end
+
+    def get_person_metric_timeline(person_id, metric_id, kwargs = {})
+      path = "#{@person}/#{person_id}/#{@metric}/#{metric_id}/#{@timeline}"
+
+      v1_request('GET', path, kwargs)
+    end
+
+# LISTS API
+
+    def create_list(list_name)
+      path = "#{@lists}"
+
+      body = {
+        "list_name": list_name
```

Comment:
list_name should be constant - https://www.rubyguides.com/2017/07/ruby-constants/